host/xtest/pkcs11_1000.c

```diff
@@ -1027,3 +1050,404 @@ static void xtest_pkcs11_test_1004(ADBG_Case_t *c)
+		{ CKA_ENCRYPT,	&(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) }, \
+		{ CKA_DECRYPT,	&(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) }, \
+		CKTEST_AES_ALLOWED_KEY(_allowed),			\
+	}
+
+#define CK_KEY_ALLOWED_AES_ENC_TEST(_label, _allowed) \
+	static CK_ATTRIBUTE _label[] = {				\
+		CKTEST_AES_KEY,						\
+		{ CKA_ENCRYPT,	&(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) }, \
+		CKTEST_AES_ALLOWED_KEY(_allowed),			\
+	}
+#define CK_KEY_ALLOWED_AES_DEC_TEST(_label, _allowed) \
+	static CK_ATTRIBUTE _label[] = {				\
+		CKTEST_AES_KEY,						\
+		{ CKA_DECRYPT,	&(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) }, \
+		CKTEST_AES_ALLOWED_KEY(_allowed),			\
+	}
+
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_only_ecb, allowed_only_aes_ecb);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_not_ecb, allowed_not_aes_ecb);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_only_cbcnopad, allowed_only_aes_cbcnopad);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_not_cbcnopad, allowed_not_aes_cbcnopad);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_only_cts, allowed_only_aes_cts);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_not_cts, allowed_not_aes_cts);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_only_ctr, allowed_only_aes_ctr);
+CK_KEY_ALLOWED_AES_TEST(cktest_aes_not_ctr, allowed_not_aes_ctr);
+
+struct cktest_allowed_test {
+	CK_ATTRIBUTE_PTR attr_key;
+	CK_ULONG attr_count;
+	CK_MECHANISM_PTR mechanism;
+};
+
+#define CKTEST_KEY_MECHA(key, mecha) {	\
+		.attr_key = key,		\
+		.attr_count = ARRAY_SIZE(key),	\
+		.mechanism = mecha,		\
+	}
+
+static const struct cktest_allowed_test cktest_allowed_valid[] = {
+	CKTEST_KEY_MECHA(cktest_aes_only_ecb, &cktest_aes_ecb_mechanism),
+	CKTEST_KEY_MECHA(cktest_aes_only_cbcnopad, &cktest_aes_cbc_mechanism),
+	CKTEST_KEY_MECHA(cktest_aes_only_cts, &cktest_aes_cts_mechanism),
+	CKTEST_KEY_MECHA(cktest_aes_only_ctr, &cktest_aes_ctr_mechanism),
+};
+
+static const struct cktest_allowed_test cktest_allowed_invalid[] = {
+	CKTEST_KEY_MECHA(cktest_aes_not_ecb, &cktest_aes_ecb_mechanism),
+	CKTEST_KEY_MECHA(cktest_aes_not_cbcnopad, &cktest_aes_cbc_mechanism),
+	CKTEST_KEY_MECHA(cktest_aes_not_cts, &cktest_aes_cts_mechanism),
+	CKTEST_KEY_MECHA(cktest_aes_not_ctr, &cktest_aes_ctr_mechanism),
+};
+
+/* Create session object and token object from a session */
+static CK_RV cipher_init_final(ADBG_Case_t *c, CK_SESSION_HANDLE session,
+				CK_ATTRIBUTE_PTR attr_key, CK_ULONG attr_count,
+				CK_MECHANISM_PTR mechanism, uint32_t mode,
+				CK_RV expected_rc)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_OBJECT_HANDLE object = CK_INVALID_HANDLE;
+
+	switch (mode) {
+	case TEE_MODE_ENCRYPT:
+	case TEE_MODE_DECRYPT:
+		break;
+	default:
+		ADBG_EXPECT_TRUE(c, false);
+	}
+
+	rv = C_CreateObject(session, attr_key, attr_count, &object);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		return rv;
+
+	if (mode == TEE_MODE_ENCRYPT)
+		rv = C_EncryptInit(session, mechanism, object);
+	if (mode == TEE_MODE_DECRYPT)
+		rv = C_DecryptInit(session, mechanism, object);
+
+	if (!ADBG_EXPECT_CK_RESULT(c, expected_rc, rv))
+		return CKR_GENERAL_ERROR;
```

Comment:
right. would need somehting like `{ rv = CKR_GENERAL_ERROR; goto out; }`

Reply:
But we don't need that since we're dealing with session objects.

Reply:
From inline comment above, this can create token object: `/* Create session object and token object from a session */`.
As `cipher_init_final()` is quite generic, I think @vesajaaskelainen is right and the function should better destroy created object, in case.

Reply:
Fair enough, I'll fix.